adapters/network.js

```diff
@@ -161,23 +162,76 @@ const httpPATCH = async (url, data, options) => {
   return clientResponse;
 };
 
-/**
- * Prepares the proxy request
- * @param {*} request
- * @returns
- */
-const prepareProxyRequest = request => {
-  const { body, method, params, endpoint } = request;
-  let { headers } = request;
-  let data;
+const getPayloadData = body => {
   let payload;
   let payloadFormat;
-  for (const [key, value] of Object.entries(body)) {
+  Object.entries(body).forEach(([key, value]) => {
     if (!_.isEmpty(value)) {
       payload = value;
       payloadFormat = key;
     }
+  });
+  return { payload, payloadFormat };
+};
+
+/**
+ * Method for stringification of query parameters
+ * To understand the use-case for this method, please take a look at the code mentioned below:
+ *
+ * @param {*} value
+ * @returns {String}
+ */
+function stringifyQueryParam(value) {
+  let stringifiedValue = `${value}`;
+  if (Array.isArray(value)) {
+    stringifiedValue = value.map(v => stringifyQueryParam(v)).join(",");
+    return `[${stringifiedValue}]`;
+  }
+  if (value && typeof value === "object") {
+    // check for value is being done to avoid null inside since typeof null = "object"
+    stringifiedValue = JSON.stringify(value);
+  }
+  return stringifiedValue;
+}
+
+/**
+ * Obtain FORM payload-format data to send the data to destination
+ *
+ * @param {Object} payload
+ * @returns {String}
+ */
+function getFormData(payload, isCustom) {
+  let data = new URLSearchParams();
+  Object.keys(payload).forEach(key => {
+    let payloadValStr = `${payload[key]}`;
+    if (typeof payload[key] === "object") {
+      payloadValStr = stringifyQueryParam(payload[key]);
+    }
+    data.append(key, payloadValStr);
+  });
+  if (!isCustom) {
```

Comment:
As this block should not be executed in production, a better way to handle this is,
`if (isCustom)` as the default value of the flag fails the check thus preventing accidental usages of this method.

Reply:
The intention is to have the encoding done by default but in-case we need something custom we'd need a way to skip the encoding for the custom-cases. Hence the condition is put that way.

Reply:
Do we have an unavoidable use case to skip this action in production or testing? If not, let's remove it.

Reply:
For `Pardot` we need to add a custom query parameter so that the request would be successful. Hence for any such custom cases, we'd probably need this kind of a flag

Reply:
It's already taken care of for Pardot destination with its own implementation of the `prepareProxyRequest` method.

How do we define this flag? 
Custom is very generic and we are loading multiple responsibilities on it.

Reply:
```javascript
const {
  prepareProxyRequest
} = require("../../../adapters/network");

...

// pardot/util.js
const { endpoint, data, method, params, headers } = prepareProxyRequest(
    request,
    true
  );

  // Modify the data
  const { payloadFormat } = getPayloadData(body);
  if (payloadFormat === "FORM") {
    data.append("format", "json");
  }
```
The `prepareProxyRequest` method returns an object with `data`(key-inside the returned object) as a string
string data-type in js doesn't have `append` method, hence we added a flag so that anyone can now operate on the URLSearchParams's object directly in order to do something very specific for that destination. Let me know if this makes sense

Reply:
> How do we define this flag?
Custom is very generic and we are loading multiple responsibilities on it.

Currently I think it is defined as a boolean value and yes, currently the definition is broad.
We can probably turn into options and apply them or think of another way of defining things in order to handle custom things, after probably having the whole picture(I mean completely testing the feature)
That way, we'd know what are exclusive to comparison and what are needed to go to production.

Reply:
It's a slippery slope to start with. It's not a good idea for a common utility to host any caller's implementation-specific behaviors.
Once the URLSearchParams object is returned, the caller can transform the data as required. BTW, we can reconstruct a URLSearchParams object from the query string using its constructor.

As far as what we are doing in the custom block, I believe they can be safely operated directly in the payload enumeration block above.

Reply:
>BTW, we can reconstruct a URLSearchParams object from the query string using its constructor.

Why would need to again do something that is already done ?


> It's not a good idea for a common utility to host any caller's implementation-specific behaviors.

The common utility is merely considering it rather than doing any computation with it. It wouldn't have any knowledge of how the caller is implementing at all. But I agree that common utility should not have any caller's implementation details
Here we are merely considering if the caller wants to do the things their way or not.


> payload enumeration block above.

Can you point out where exactly it is ?

Reply:
Made the needed changes. Resolving this.